java-squid/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -0,0 +1,750 @@
+      for (CaseGroupTree caseGroupTree : Lists.reverse(switchStatementTree.cases())) {
+        build(caseGroupTree.body());
+        switches.getLast().successors.add(currentBlock);
+        if (caseGroupTree != firstCase) {
+          // No block predecessing the first case group.
+          currentBlock = createBlock(currentBlock);
+        }
+      }
+    }
+    breakTargets.removeLast();
+    // process condition
+    currentBlock = switches.removeLast();
+  }
+
+  private void buildBreakStatement(BreakStatementTree tree) {
+    if (tree.label() == null) {
+      if (breakTargets.isEmpty()) {
+        throw new IllegalStateException("'break' statement not in loop or switch statement");
+      }
+      currentBlock = createUnconditionalJump(tree, breakTargets.getLast());
+    } else {
+      currentBlock = createUnconditionalJump(tree, null);
+      gotos.add(currentBlock);
+    }
+  }
+
+  private void buildContinueStatement(ContinueStatementTree tree) {
+    if (tree.label() == null) {
+      if (continueTargets.isEmpty()) {
+        throw new IllegalStateException("'continue' statement not in loop or switch statement");
+      }
+      currentBlock = createUnconditionalJump(tree, continueTargets.getLast());
+    } else {
+      currentBlock = createUnconditionalJump(tree, null);
+      gotos.add(currentBlock);
+    }
+  }
+
+  private void buildWhileStatement(WhileStatementTree tree) {
+    WhileStatementTree s = tree;
+    Block falseBranch = currentBlock;
+    Block loopback = createBlock();
+    // process body
+    currentBlock = createBlock(loopback);
+    continueTargets.addLast(loopback);
+    breakTargets.addLast(falseBranch);
+    build(s.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    Block bodyBlock = currentBlock;
+    // process condition
+    currentBlock = createBranch(s, bodyBlock, falseBranch);
+    buildCondition(s.condition(), bodyBlock, falseBranch);
+    loopback.successors.add(currentBlock);
+    currentBlock = createBlock(currentBlock);
+  }
+
+  private void buildDoWhileStatement(DoWhileStatementTree tree) {
+    DoWhileStatementTree s = tree;
+    Block falseBranch = currentBlock;
+    Block loopback = createBlock();
+    // process condition
+    currentBlock = createBranch(s, loopback, falseBranch);
+    buildCondition(s.condition(), loopback, falseBranch);
+    // process body
+    currentBlock = createBlock(currentBlock);
+    continueTargets.addLast(loopback);
+    breakTargets.addLast(falseBranch);
+    build(s.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    loopback.successors.add(currentBlock);
+    currentBlock = createBlock(currentBlock);
+  }
+
+  private void buildForEachStatement(ForEachStatement tree) {
+    ForEachStatement s = tree;
+    // TODO(npe) One solution is to create a forstatement node depending on type of expression (iterable or array) and build CFG from
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)